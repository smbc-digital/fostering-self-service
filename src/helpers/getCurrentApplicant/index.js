import { Applicant, SECOND_APPLICANT_ROUTE } from 'constants'

const getCurrentApplicant = match => {
    if (match.params && match.params[0] === SECOND_APPLICANT_ROUTE) {
        return Applicant.SecondApplicant
    }

    return Applicant.FirstApplicant
}

export default getCurrentApplicant