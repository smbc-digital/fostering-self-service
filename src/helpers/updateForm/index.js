import { fetchWithTimeout } from '../index'

const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

const StageName = {
    HomeVisit: 0,
    Application: 1
}

const HomeVisitFormName = {
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguagesSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 8
}

const ApplicationFormName = {
    References: 0,
    GpDetails: 1,
    AddressHistory: 2
}

const getHomeVisitUpdateEndpoint = form => {
    switch (form) {
        case HomeVisitFormName.TellUsAboutYourself:
            return '/fostering/home-visit/about-yourself'
        case HomeVisitFormName.YourEmploymentDetails:
            return '/fostering/home-visit/your-employment-details'
        case HomeVisitFormName.LanguagesSpokenInYourHome:
            return '/fostering/home-visit/languages-spoken-in-your-home'
        case HomeVisitFormName.YourFosteringHistory:
            return '/fostering/home-visit/your-fostering-history'
        case HomeVisitFormName.YourPartnership:
            return '/fostering/home-visit/partnership-status'
        case HomeVisitFormName.TellUsAboutYourInterestInFostering:
            return '/fostering/home-visit/interest-in-fostering'
        case HomeVisitFormName.YourHealth:
            return '/fostering/home-visit/about-your-health'
        case HomeVisitFormName.YourHousehold:
            return '/fostering/home-visit/household'
        case HomeVisitFormName.ChildrenLivingAwayFromYourHome:
            return '/fostering/home-visit/children-living-away-from-home'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getApplicationUpdateEndpoint = form => {
    switch (form) {
        case ApplicationFormName.References:
            return '/fostering/application/references'
        case ApplicationFormName.GpDetails:
            return '/fostering/application/gp-details'
        default:
            throw new Error('No matching endpoint for given form.')
    }
}

const getFormStageEndpoint = stage => {
    switch (stage) {
        case StageName.HomeVisit: 
            return '/fostering/home-visit'
        case StageName.Application:
            return '/fostering/application'
        default:
            throw new Error('No matching endpoint for given stage.')
    }
}

const reduceProperties = object => Object.keys(object).reduce((acc, property) => {
    return {
        ...acc,
        [property]: object[property].value
    }
}, {})

const updateForm = async (endpoint, formData) => {
    const parsedFormData = parseFormData(formData)

    const response = await fetchWithTimeout(endpoint, {
        method: 'PATCH',
        credentials: 'include',
        body: JSON.stringify(parsedFormData),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
    }, 30000)

    if (!response.ok) {
        throw Error(response.statusText)
    }

    return await response.json()
}

const updateFormStatus = ({ form, stage, currentStatus, setStatus }) => {
    if (currentStatus === TaskStatus.None) {
        fetchWithTimeout(`${getFormStageEndpoint(stage)}/status`,
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                }),
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            }, 30000)

        setStatus(TaskStatus.NotCompleted)
    }
}

const parseFormData = ({ firstApplicant, secondApplicant, familyReference, firstPersonalReference, secondPersonalReference, ...formData }) => {
    let parsedObject = reduceProperties(formData)

    if (firstApplicant) {
        parsedObject.firstApplicant = reduceProperties(firstApplicant)
    }

    if (secondApplicant) {
        parsedObject.secondApplicant = reduceProperties(secondApplicant)
    }

    if (familyReference) {
        parsedObject.familyReference = reduceProperties(familyReference)
        parsedObject.familyReference.address = familyReference.address.value
    }

    if (firstPersonalReference) {
        parsedObject.firstPersonalReference = reduceProperties(firstPersonalReference)
        parsedObject.firstPersonalReference.address = firstPersonalReference.address.value
    }

    if (secondPersonalReference) {
        parsedObject.secondPersonalReference = reduceProperties(secondPersonalReference)
        parsedObject.secondPersonalReference.address = secondPersonalReference.address.value
    }

    return parsedObject
}

const updateApplicationForm = async (form, formData) => {
    const endpoint = getApplicationUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

const updateHomeVisitForm = async (form, formData) => {
    const endpoint = getHomeVisitUpdateEndpoint(form)

    return await updateForm(endpoint, formData)
}

export {
    TaskStatus,
    StageName,
    HomeVisitFormName,
    ApplicationFormName,
    updateFormStatus,
    parseFormData,
    updateApplicationForm,
    updateHomeVisitForm
}