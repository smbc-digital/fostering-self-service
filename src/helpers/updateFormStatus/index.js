export const TaskStatus = {
    None: 0,
    Completed: 1,
    NotCompleted: 2,
    CantStart: 3
}

export const FormName =
{
    TellUsAboutYourself: 0,
    YourEmploymentDetails: 1,
    LanguageSpokenInYourHome: 2,
    YourPartnership: 3,
    YourFosteringHistory: 4,
    YourHealth: 5,
    TellUsAboutYourInterestInFostering: 6,
    YourHousehold: 7,
    ChildrenLivingAwayFromYourHome: 7
}

export default function (form, currentStatus, setStatus) {
    if (currentStatus === TaskStatus.None) {
        fetch('/front-end-endpoint-update-status',
            {
                method: 'PATCH',
                credentials: 'include',
                body: JSON.stringify({
                    status: TaskStatus.NotCompleted,
                    form
                })
            })

        setStatus(TaskStatus.NotCompleted)
    }
}