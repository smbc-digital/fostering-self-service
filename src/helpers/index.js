import fetchWithTimeout from './fetchWithTimeout'
import { getPageRoute } from './pagehelper'
import getCurrentApplicant from './getCurrentApplicant'
import { updateFormStatus, updateForm, FormName, TaskStatus } from './updateForm'

export {
    fetchWithTimeout,
    getPageRoute,
    getCurrentApplicant,
    updateFormStatus,
    updateForm,
    FormName,
    TaskStatus
}