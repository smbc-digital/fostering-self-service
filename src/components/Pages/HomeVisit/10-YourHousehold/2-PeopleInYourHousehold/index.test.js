import { React, mount, useContextMock, renderer } from 'helpers/SetupTest'
import PeopleInYourHousehold from './index'
import * as helpers from 'helpers'

describe('PeopleInYourHousehold', () => {
    const mockOnChange = jest.fn()

    beforeEach(() => {
        useContextMock.mockReturnValue({
            onChange: mockOnChange,
            otherPeopleInYourHousehold: {
                value: [{'IsDobValid': false, 'dateOfBirth': '2016-02-20', 'firstName': 'test'}],
                isValid: false
            },
        })
    })

    it('Should call history.push on submit', () => {
        // Arrange
        const history = {
            push: jest.fn()
        }

        // Act
        const wrapper = mount(<PeopleInYourHousehold history={history} />)
        wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(helpers.getPageRoute(16))
    })

    it('Should call onChange with new values', () => {
        // Act
        const wrapper = mount(<PeopleInYourHousehold history={history} />)
        wrapper.find('input').at(1).simulate('change', { target: { value: 'test', name: 'firstName' } })

        // Assert
        expect(mockOnChange).toHaveBeenCalled()
    })

    it('Should call onChange with IsDobValid value', () => {
        // Act
        const wrapper = mount(<PeopleInYourHousehold history={history} />)
        mockOnChange.mockReset()
        wrapper.find('input[name="year"]').simulate('change', { target: { value: '2015', name: 'year' } })

        // Assert
        expect(mockOnChange).toHaveBeenCalledWith({'target': {'name': 'otherPeopleInYourHousehold', 'value': [{'IsDobValid': true, 'dateOfBirth': '2015-2-20', 'firstName': 'test'}]}}, true)
    })

    describe('snapshot', () => {
        it('renders correctly', () => {
            const tree = renderer
                .create(<PeopleInYourHousehold history={{}}/>)
                .toJSON()

            expect(tree).toMatchSnapshot()
        })
    })
})