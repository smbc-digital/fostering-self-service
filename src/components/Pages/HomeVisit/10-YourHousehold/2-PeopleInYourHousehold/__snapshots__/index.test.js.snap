// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleInYourHousehold snapshot renders correctly 1`] = `
Array [
  <h1>
    Your fostering journey
  </h1>,
  <h2>
    Your household
  </h2>,
  <h3>
    Tell us who lives in your home
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <label
      className="optional"
      htmlFor="firstName"
    >
      First name
    </label>
    <input
      aria-label={undefined}
      aria-live={undefined}
      className=""
      id="firstName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="firstName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value="test"
    />
    <label
      className="optional"
      htmlFor="lastName"
    >
      Last name
    </label>
    <input
      aria-label={undefined}
      aria-live={undefined}
      className=""
      id="lastName"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="lastName"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value={undefined}
    />
    <label
      className="optional"
      htmlFor="gender"
    >
      Gender
    </label>
    <input
      aria-label={undefined}
      aria-live={undefined}
      className=""
      id="gender"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="gender"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value={undefined}
    />
    <label
      className="optional"
      htmlFor="relationshipToYou"
    >
      Relationship to you
    </label>
    <input
      aria-label={undefined}
      aria-live={undefined}
      className=""
      id="relationshipToYou"
      inputMode={undefined}
      maxLength={60}
      min={undefined}
      name="relationshipToYou"
      onChange={[Function]}
      onKeyDown={[Function]}
      pattern={undefined}
      type="text"
      value={undefined}
    />
    <h3
      className="optional no-margin-bottom"
    >
      Date of birth
    </h3>
    <fieldset
      className="input-memorable-date"
    >
      <p>
        For example, 31 3 1980
      </p>
      <div>
        <label>
          Day
        </label>
        <input
          aria-label="day"
          className=""
          name="day"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value="20"
        />
      </div>
      <div>
        <label>
          Month
        </label>
        <input
          aria-label="month"
          className=""
          name="month"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value="2"
        />
      </div>
      <div>
        <label>
          Year
        </label>
        <input
          aria-label="year"
          className=""
          name="year"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value="2016"
        />
      </div>
    </fieldset>
    <a
      className="anchor-with-icon"
      href="#"
      onClick={[Function]}
    >
       
      <i
        className="fa fa-plus-square fa-lg"
      />
      Add another person
    </a>
    <button
      className="button-primary"
      disabled={false}
      id={undefined}
      onClick={undefined}
      type="submit"
    >
      Next step
    </button>
  </form>,
  <p>
    <a
      href="#"
      onClick={[Function]}
    >
      Previous
    </a>
  </p>,
]
`;
